src/Chapter3.hs

```diff
@@ -1007,6 +1135,145 @@ contestants, and write a function that decides the outcome of a fight!
+  | Defeat deriving Show
+
+data FightKnight = FightKnight
+    { fightKnightHealth  :: FightHealth
+    , fightKnightAttack :: FightAttack
+    , fightKnightDefence :: FightDefense
+    , fightKnightActions :: [FightAction]
+    } deriving Show
+
+data FightMonster = FightMonster
+    { fightMonsterHealth  :: FightHealth
+    , fightMonsterAttack :: FightAttack
+    , fightMonsterActions :: [FightAction]
+    } deriving Show
+
+class CharacterClass a where
+    getHealth :: a -> FightHealth
+    getAttack :: a -> FightAttack
+
+instance CharacterClass FightKnight where
+  getHealth = fightKnightHealth
+  getAttack = fightKnightAttack
+
+instance CharacterClass FightMonster where
+  getHealth = fightMonsterHealth
+  getAttack = fightMonsterAttack
+
+class CharacterDefenceClass a where
+    getDefence :: a -> FightDefense
+
+instance CharacterDefenceClass FightKnight where
+  getDefence = fightKnightDefence
+
+processRoundAction :: FightAction -> FightKnight -> FightMonster -> FightResult
+processRoundAction action knight monster =
+  case action of
+    MonsterAttack ->
+      let
+        FightHealth health = getHealth knight
+        FightDefense defence = getDefence knight
+        FightAttack attack = getAttack monster
+        newHealth = health - max 0 (defence - defence)
+      in
+        if newHealth >= 0
+          then Round (knight { fightKnightHealth = FightHealth newHealth }) monster
+          else Defeat
+
+    MonsterRun ->
+      Victory
+
+    KnightCastSpell cast ->
+      let
+        FightDefense defence = getDefence knight
+        newDefense = defence + cast
+        newKnight = knight { fightKnightDefence = FightDefense newDefense}
+      in
+        Round newKnight monster
+
+    KnightDrinkPotion potion ->
+      let
+        FightHealth health = getHealth knight
+        newHealth = health + potion
+        newKnight = knight { fightKnightHealth = FightHealth newHealth }
+      in
+        Round newKnight monster
+
+    KnightAttack ->
+      let
+        FightHealth health = getHealth monster
+        FightAttack damage = getAttack knight
+        newHealth = health - damage
+      in
+        if newHealth >= 0
+          then Round knight (monster { fightMonsterHealth = FightHealth newHealth })
+          else Victory
+
+processFightActions :: FightResult -> [FightAction] -> FightResult
+processFightActions Victory _ = Victory
+processFightActions Defeat _ = Defeat
+processFightActions Wait _ = Wait
+processFightActions (Round knight monster) [] =
+    let newActions = createFightActions knight monster in
+      if length newActions == 0 then Wait else
+        processFightActions (Round knight monster) newActions
+
+processFightActions (Round knight monster) (action:actions) =
+    processFightActions (processRoundAction action knight monster) actions
+
+createFightActions :: FightKnight -> FightMonster -> [FightAction]
+createFightActions knight monster =
+  fightKnightActions knight ++ fightMonsterActions monster
```

Comment:
Oh, Interesting!
Yeah, wouldn't in this case, knight perform all its actions before monster even starts the fight?

Reply:
I've implemented as I've understood the following >_< :
```
Each fighter starts with some list of actions they can do, performs them in sequence one after another...
```
Now I've got that action list should be look like:

```
[KnightAttack, MonsterAttack, KnightDrinkPotion Int, .. ]
```

Reply:
Yeap, I see that this is just another understanding of the task, no worries 🙂